### PR TITLE
Allow all X.509 Signature Algorithms, not just RSA-SHA256

### PR DESCRIPTION
**Pull request by vinarmani**
There is a bug where, essentially, SHA256 is the only signature algorithm that will be checked for the SSL certificate chain. I have changed it so that it will use the correct algorithm for each certificate in the chain based on the decoded certificate information.

While this bug exists, valid payment requests signed, for instance, with the SHA384 algorithm will be seen as invalid by the wallet.

### CODE REVIEW

**Comment by vinarmani** [comment]
Don't want to include new BitPay changes

**Comment by vinarmani** [comment]
The BitPay changes are positive and don't seem to conflict.

File: lib/browser.js [hunk 1]
```diff
@@ -196,8 +196,10 @@ PaymentProtocol.verifyCertChain = function(chain, sigHashAlg) {
 
     // Verify current Certificate signature
     if (sigHashAlg !== 'none') {
+      let sigAlgId = c.signatureAlgorithm.algorithm.join('.')
```
Comment (brendoncoin) [inline lib/browser.js]: Replace let with var. The Wallet runs on old devices that do not support let.

File: package.json [hunk 2]
```diff
@@ -49,20 +49,22 @@
...
     "bs58": "^4.0.1",
-    "jsrsasign": "^8.0.4",
-    "protobufjs": "=5.0.1"
+    "jsrsasign": "^8.0.12",
+    "protobufjs": "=5.0.3"
+  },
+  "peerDependencies": {
+    "bitcore-lib": "=0.16.0"
   },
   "devDependencies": {
```
Comment (brendoncoin) [inline package.json:60]: Do the latest versions of all of these dependencies support the old devices we need to run on?

File: README.md [hunk 3]
```diff
@@ -9,19 +9,15 @@ A module for [bitcore](https://github.com/bitpay/bitcore) that implements [Payme
...
 npm install bitcore-payment-protocol
 ```
 
-```sh
-bower install bitcore-lib
-bower install bitcore-payment-protocol
-```
 
-There are many examples of how to use it on the developer guide [section for payment protocol](https://bitcore.io/api/paypro). For example, the following code would verify a payment request:
+There are many examples of how to use it on the developer guide [section for payment protocol](https://github.com/vinarmani/bitcore-payment-protocol/blob/master/docs/index.md). For example, the following code would verify a payment request:
```
Comment (brendoncoin) [inline README.md]: This should link to bitcoin.com.

**Review by brendoncoin — CHANGES_REQUESTED** [review]

File: README.md (hunk 3 above)
Comment (vinarmani) [reply README.md]: Done

File: package.json (hunk 2 above)
Comment (vinarmani) [reply package.json:60]: The latest versions do not appear to be necessary for proper functionality. They were updated by the BitPay devs as what appears to be housekeeping and not to enable any new features or fix any bugs.

File: lib/browser.js [hunk 4]
```diff
@@ -196,8 +196,10 @@ PaymentProtocol.verifyCertChain = function(chain, sigHashAlg) {
 
     // Verify current Certificate signature
     if (sigHashAlg !== 'none') {
+      var sigAlgId = c.signatureAlgorithm.algorithm.join('.')
+      var sigAlg = PaymentProtocol.getAlgorithm(sigAlgId).split('_')
       var jsrsaSig = new KJUR.crypto.Signature({
-        alg: sigHashAlg + 'withRSA',
+        alg: sigAlg[1].toLowerCase() + 'withRSA',
```
Comment (jbdtky) [inline lib/browser.js]: getAlgorithm can return 'RSA'. So the split will return `'1.2.840.113549.1.1.1': 'RSA'` so sigAlg[1] will fail.
https://github.com/bitpay/bitcore-payment-protocol/blob/dc7e7fcdab8e06f212c820f47d6961dd85ec1265/lib/common.js#L48
Can you make it safe?

File: package.json [hunk 5]
```diff
@@ -52,17 +52,17 @@
     "asn1.js": "=4.9.1",
     "asn1.js-rfc3280": "=4.0.0",
     "asn1.js-rfc5280": "=2.0.0",
-    "bitcore-lib": "^0.15.0",
+    "bitcore-lib": "0.14.0",
```
Comment (jbdtky) [inline package.json:55]: Did you try to use the 0.15.0? If yes, what was the issue?

File: .travis.yml [hunk 6]
```diff
@@ -1,14 +1,20 @@
 language: node_js
+sudo: false
```
Comment (jbdtky) [inline .travis.yml]: sudo is already defined below. I think you should remove this one.

**Review by jbdtky — CHANGES_REQUESTED** [review]
@vinarmani, can you have a look at my comments?

File: lib/browser.js (hunk 4 above)
Comment (vinarmani) [reply lib/browser.js]: Alright, done and successfully tested in a new browser build of the Bitcoin.com wallet using RSA-SHA256 and RSA-SHA384 certs in the wild.

File: package.json (hunk 5 above)
Comment (vinarmani) [reply package.json:55]: There was no issue. I changed it to 0.14.0 so there weren't conflicts with the existing version. Either version works fine. It's actually a peer dependency as I don't believe it is actually used in the code.

File: .travis.yml (hunk 6 above)
Comment (vinarmani) [reply .travis.yml]: done

File: lib/browser.js [hunk 7]
```diff
@@ -198,19 +198,25 @@ PaymentProtocol.verifyCertChain = function(chain, sigHashAlg) {
...
-
-      // Get the raw DER TBSCertificate
-      // from the DER Certificate:
-      var tbs = PaymentProtocol.getTBSCertificate(data, sig);
-
-      jsrsaSig.updateHex(tbs.toString('hex'));
-
-      sigVerified = jsrsaSig.verify(sig.toString('hex'));
+      // If the signature algorithm is plain RSA, the cert is invalid
+      console.log(sigAlg)
```
Comment (jbdtky) [inline lib/browser.js]: Please, remove your `console.log`.

**Review by jbdtky — CHANGES_REQUESTED** [review]

File: lib/browser.js (hunk 7 above)
Comment (vinarmani) [reply lib/browser.js]: That was already removed and committed days ago
https://github.com/Bitcoin-com/bitcore-payment-protocol/pull/1/commits/d37211de345f74b2b0b24b43eebae990ce1015ee

**Pull request merged**